src/gpl/src/nesterovBase.cpp

```diff
@@ -3003,69 +3053,136 @@ size_t NesterovBaseCommon::createCbkGCell(odb::dbInst* db_inst)
 void NesterovBase::destroyCbkGCell(odb::dbInst* db_inst)
 {
-  auto db_it = db_inst_index_map_.find(db_inst);
-  if (db_it != db_inst_index_map_.end()) {
+  debugPrint(log_, GPL, "callbacks", 2, "NesterovBase::destroyGCel");
+  auto db_it = db_inst_to_nb_index_map_.find(db_inst);
+  if (db_it != db_inst_to_nb_index_map_.end()) {
     size_t last_index = nb_gcells_.size() - 1;
     size_t gcell_index = db_it->second;
 
     GCellHandle& handle = nb_gcells_[gcell_index];
-    if (handle.isNesterovBaseCommon()) {
-      nbc_->destroyCbkGCell(handle.getIndex());
-    } else {
-      destroyFillerGCell(handle.getIndex());
-    }
+    // TODO delete fillers
+    bool is_filler = handle->isFiller();
 
     if (gcell_index != last_index) {
       std::swap(nb_gcells_[gcell_index], nb_gcells_[last_index]);
     }
     swapAndPopParallelVectors(gcell_index, last_index);
     nb_gcells_.pop_back();
-    db_inst_index_map_.erase(db_it);
+    db_inst_to_nb_index_map_.erase(db_it);
+
+    if (gcell_index != last_index) {
+      if (!nb_gcells_[gcell_index]->isFiller()) {
+        odb::dbInst* swapped_inst
+            = nb_gcells_[gcell_index]->insts()[0]->dbInst();
+        db_inst_to_nb_index_map_.erase(swapped_inst);
+        db_inst_to_nb_index_map_[swapped_inst] = gcell_index;
+      }
+    } else {
+      db_inst_to_nb_index_map_.erase(db_inst);
+    }
+
+    if (!is_filler) {
+      std::pair<odb::dbInst*, size_t> replacer = nbc_->destroyCbkGCell(db_inst);
+
+      if (replacer.first != nullptr) {
+        auto it = db_inst_to_nb_index_map_.find(replacer.first);
+        if (it != db_inst_to_nb_index_map_.end()) {
+          nb_gcells_[it->second].updateIndex(replacer.second);
+        } else {
+          debugPrint(log_,
+                     GPL,
+                     "callbacks",
+                     1,
+                     "warn replacer dbInst {} not found in NB map!",
+                     replacer.first->getName());
+        }
+      }
+    } else {
+      debugPrint(log_,
+                 GPL,
+                 "callbacks",
+                 1,
+                 "warning: trying to destroy filler gcell!");
+      destroyFillerGCell(handle.getIndex());
+    }
+
+  } else {
+    debugPrint(
+        log_,
+        GPL,
+        "callbacks",
+        1,
+        "warning: db_inst not found in db_inst_index_map_ for instance: {}",
+        db_inst->getName());
   }
 }
 
-void NesterovBaseCommon::destroyCbkGCell(size_t index_remove)
+std::pair<odb::dbInst*, size_t> NesterovBaseCommon::destroyCbkGCell(
+    odb::dbInst* db_inst)
 {
-  size_t last_index = gCellStor_.size() - 1;
-  if (index_remove > last_index) {
-    return;
+  auto it = db_inst_to_nbc_index_map_.find(db_inst);
+  if (it == db_inst_to_nbc_index_map_.end()) {
+    log_->error(GPL,
+                307,
+                "db_inst not found in db_inst_to_NBC_index_map_ when trying to "
+                "destroy GCell on NBC");
+    return {};
```

Comment:
You don't need this return here. The `Logger::error` should be enough to stop the execution of this function.